src/jdk.jartool/share/classes/sun/security/tools/jarsigner/Main.java

```diff
@@ -1128,6 +1133,123 @@ void verifyJar(String jarName)
         }
     }
 
+    private void crossCheckEntries(String jarName) throws Exception {
+        List<String> locEntries = new ArrayList<>();
+        List<String> cenEntries;
+
+        try (JarFile jarFile = new JarFile(jarName);
+             JarInputStream jis = new JarInputStream(
+                     Files.newInputStream(Path.of(jarName)))) {
+
+            Manifest cenManifest = jarFile.getManifest();
+            Manifest locManifest = jis.getManifest();
+            if (!compareManifest(cenManifest, locManifest)) {
+                return;
+            }
+
+            JarEntry locEntry;
+            while ((locEntry = jis.getNextJarEntry()) != null) {
+                String entryName = locEntry.getName();
+                locEntries.add(entryName);
+
+                JarEntry cenEntry = jarFile.getJarEntry(entryName);
+                if (cenEntry == null) {
+                    crossChkWarnings.add(String.format(rb.getString(
+                            "Entry.missing.in.JarFile.1"), entryName));
```

Comment:
Would it be more precise to change this warning to "Entry %s is present when reading via JarInputStream but missing when reading via JarFile"?

Reply:
Updated this warning as suggested.